mempool/mempool.go

```diff
@@ -924,313 +924,39 @@ func (mp *TxPool) validateReplacement(tx *btcutil.Tx,
-	// double spend, unless those transactions signal for RBF. This check is
-	// intended to be quick and therefore only detects double spends within
-	// the transaction pool itself. The transaction could still be double
-	// spending coins from the main chain at this point. There is a more
-	// in-depth check that happens later after fetching the referenced
-	// transaction inputs from the main chain which examines the actual
-	// spend data and prevents double spends.
-	isReplacement, err := mp.checkPoolDoubleSpend(tx)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	// Fetch all of the unspent transaction outputs referenced by the inputs
-	// to this transaction.  This function also attempts to fetch the
-	// transaction itself to be used for detecting a duplicate transaction
-	// without needing to do a separate lookup.
-	utxoView, err := mp.fetchInputUtxos(tx)
-	if err != nil {
-		if cerr, ok := err.(blockchain.RuleError); ok {
-			return nil, nil, chainRuleError(cerr)
-		}
-		return nil, nil, err
-	}
-
-	// Don't allow the transaction if it exists in the main chain and is
-	// already fully spent.
-	prevOut := wire.OutPoint{Hash: *txHash}
-	for txOutIdx := range tx.MsgTx().TxOut {
-		prevOut.Index = uint32(txOutIdx)
-		entry := utxoView.LookupEntry(prevOut)
-		if entry != nil && !entry.IsSpent() {
-			return nil, nil, txRuleError(wire.RejectDuplicate,
-				"transaction already exists")
-		}
-		utxoView.RemoveEntry(prevOut)
-	}
-
-	// Transaction is an orphan if any of the referenced transaction outputs
-	// don't exist or are already spent.  Adding orphans to the orphan pool
-	// is not handled by this function, and the caller should use
-	// maybeAddOrphan if this behavior is desired.
-	var missingParents []*chainhash.Hash
-	for outpoint, entry := range utxoView.Entries() {
-		if entry == nil || entry.IsSpent() {
-			// Must make a copy of the hash here since the iterator
-			// is replaced and taking its address directly would
-			// result in all the entries pointing to the same
-			// memory location and thus all be the final hash.
-			hashCopy := outpoint.Hash
-			missingParents = append(missingParents, &hashCopy)
-		}
-	}
-	if len(missingParents) > 0 {
-		return missingParents, nil, nil
-	}
-
-	// Don't allow the transaction into the mempool unless its sequence
-	// lock is active, meaning that it'll be allowed into the next block
-	// with respect to its defined relative lock times.
-	sequenceLock, err := mp.cfg.CalcSequenceLock(tx, utxoView)
-	if err != nil {
-		if cerr, ok := err.(blockchain.RuleError); ok {
-			return nil, nil, chainRuleError(cerr)
-		}
-		return nil, nil, err
-	}
-	if !blockchain.SequenceLockActive(sequenceLock, nextBlockHeight,
-		medianTimePast) {
-		return nil, nil, txRuleError(wire.RejectNonstandard,
-			"transaction's sequence locks on inputs not met")
-	}
+func (mp *TxPool) maybeAcceptTransaction(tx *btcutil.Tx, isNew, rateLimit,
+	rejectDupOrphans bool) ([]*chainhash.Hash, *TxDesc, error) {
```

Comment:
Def more that we should do here in the future, but +1 to chopping up the routine a bit.